Qualcommax: IPQ807x: Add support for TP-Link Deco X80-5G

professor-jonny opened the pull request:
Qualcommax: IPQ807x: Add support for TP-Link Deco X80-5G
The TP-link Deco X80-5G is a AX WIFI router with a single 1G (LAN) and 1 2.5G (WAN) port with a built in Quectel 5G modem.

Specifications:

Architecture	ARMv8-A (aarch64 Cortex A53, 4 cores)
Vendor	Qualcomm
Bootloader	U-Boot (2016)
System-On-Chip	Qualcomm IPQ8074A (Networking Pro 1210 platform)
CPU/Speed	2.2GHZ
Flash-Chip	ESMT F59D1G81MB-45TI 1G-bit NAND
Flash MB	128MB (1Gbit)
RAM-chip	2* ESMT M15T4G16256A-DEBG2G 2Gb DDR3L (32bit)
RAM MB	512MB (4Gbit)
WLan System-On-Chip (5g)	Qualcomm QCN5054
WLan Front end modules (5g)	8* Qorvo QPF4588
WLAN (5G)	A/N/AC/AX
WLAN Antenna (int)	8* (4 shared with 2.5G)
WLan System-On-Chip (2.4G)	Qualcomm QCN5024
WLan Front end modules (2.4G)	4* Skyworks SKY85340-11
WLAN (2.4G)	B/G/N/AX
WLAN Antenna (int)	4* (shared with 5g)
Eithernet-phy (1Gbit)	Qualcomm (Atheros) AR8033-AL1A
Ethernet (1Gbit)	1*
Eithernet-phy (2.5Gbit)	Qualcomm QCA8081
Ethernet (2.5Gbit)	1*
Switch	Embedded in the SOC?
Serial	1.8v 8n1 (testpoints)
Modem System-On-Chip	Qualcomm 5G RG50xQ (Snapdragon X55 platform)
5G modem capability	5000Mbps down 900Mbps up
5G modem bands	N1/N3/N5/N7/N8/N20/N28/N38/N40/N41/N77/N78/N79
4G modem capability	4.5G LTE-Advanced Pro (Cat20 down 2Gbps) (Cat18 up 200Mbps)
4G modem FDD bands	B1/B3/B5/B7/B8/B18/B19/B20/B26/B28
4G modem TDD bands	B34/B38/B39/B40/B41/B42/B43
Modem antenna (int)	8* (5G/4G use)
Modem antenna (ext)	2* SMA connectors (5G/4G use)
SIM type	1* Nano SIM
Telephony	1* POTS RJ11
(see wiki for more info:)
https://openwrt.org/inbox/toh/tp-link/x80-5g_v1

        Notes:
Installation Instructions:

Between antennas 7 and 8 on the PCB with the RGB harness there are three test pads labelled TP1 TP2 and TP3 connections are as follows:
  |TP2|-- RX
  |TP1|-- TX
  |TP3|-- GND

RX requires an external pullup to operate somewhere around 5 K ohm but your luck may vary the Uart is 1.8v

Set a static IP and set up a tftpserver and terminal.
power the router and quickly type the magic string “tpl” and press enter to break into u-boot
in the shell set the environment variables to enable tftp booting
setenv ipaddr (routerIP)
setenv serverip (server IP)
load you initramfs:
tftpboot 0x44000000 (serverIP):openwrt-qualcommax-ipq807x-tplink_x80-5g-initramfs-uImage.itb
boot your initramfs
bootm
upload your factory image to /tmp
format and install the factory image:
ubiformat /dev/mtd12 -y -f /tmp/openwrt-qualcommax-ipq807x-tplink_x80-5g-squashfs-factory.ubi

Note as this device swaps root partitions upon update your boot partition may be set as mtd13

Update script to mount factory tplink partition's for MAC recovery and device data.
Capture Labelmac from Factory_data partition.
Patch LAN from Labelmac.
Patch ART from Labelmac.
Set GPIO to make modem operational from boot.

GPIO fan tables work like the  stock device with three fan maps high active idle with a top fan speed of 6800 RPM.

High temp 70 deg speeds 6-8
Active temp 50 deg speeds 2-5
low temp 25 degree speed 0-1

Signed-off-by: Jonathan Brophy <professor_jonny@hotmail.com>

professor-jonny commented:
Of note the led script currently the script works from shell but it does not seem to run from preinit I don't know why.

professor-jonny commented:
The board file is submitted pending review
https://github.com/openwrt/firmware_qca-wireless/pull/49

musashino205 commented:
Please drop merge commits.

musashino205 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-x80-5g.dts:
```diff
@@ -0,0 +1,476 @@
+		};
+	};
+
+		antennas {
```
wrong indent

musashino205 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-x80-5g.dts:
```diff
@@ -0,0 +1,476 @@
+		pinctrl-0 = <&button_pins>;
+		pinctrl-names = "default";
+
+		button@1 {
```
`button-reset` (`gpio-keys` doesn't use `reg` property and `@` shoudn't be used)

musashino205 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-x80-5g.dts:
```diff
@@ -0,0 +1,476 @@
+		//to mimic the oem default behaviour green and red are switched on from boot to indicate yellow
+		//to show the device is booting behaviour can be user defined in procd to override this.
+
+		led_system_green: green {
+			gpios = <&tlmm 50 GPIO_ACTIVE_HIGH>; //green led component of RGB led
+			color = <LED_COLOR_ID_GREEN>;
+			default-state = "on";
+		};
+
+		led_system_red: red {
+			gpios = <&tlmm 51 GPIO_ACTIVE_HIGH>; //red led component of RGB led
+			color = <LED_COLOR_ID_RED>;
+			default-state = "on";
+		};
+
+		led_system_blue: blue {
+			gpios = <&tlmm 52 GPIO_ACTIVE_HIGH>; //blue led component of RGB
+			color = <LED_COLOR_ID_BLUE>;
+		};
```
Please add missing `function` properties.

musashino205 replied:
And please use `led-*` aliases in `aliases {}` node for status LED handling of OpenWrt instead of user-space script.

professor-jonny replied:
The LED's don't have a set assigned function they have multiple as they are not singular LED's it is a multifunction RGB LED.
There is no monochromatic LED support in OpenWrt and no support for virtual LED's' by mixing colors to be able to assign individual functions to individual colors.
I don't know how to proceed, and other targets has left them blank as I have.

musashino205 replied:
`function` property is used for LED name `<color>:<function>` (example: `green:status`) in **/sys/class/leds/**.

IMO:

```
aliases {
	led-boot = &led_system_green;
	led-falsafe = &led_system_red;
	led-running = &led_system_green;
	led-upgrade = &led_system_green;
	...
}
...
led_system_green: led-0 {
	gpios = <&tlmm 50 GPIO_ACTIVE_HIGH>;
	color = <LED_COLOR_ID_GREEN>;
	function = LED_FUNCTION_STATUS;
};

led_system_red: led-1 {
	gpios = <&tlmm 51 GPIO_ACTIVE_HIGH>;
	color = <LED_COLOR_ID_RED>;
	function = LED_FUNCTION_STATUS;
	default-state = "off";
};

led-2 {
	gpios = <&tlmm 52 GPIO_ACTIVE_HIGH>;
	color = <LED_COLOR_ID_BLUE>;
	function = LED_FUNCTION_STATUS;
	default-state = "off";
};
```

professor-jonny replied:
how do I assign an alias to mimic the oem colours ? as power is yellow in the device which is a mix of red and green led's

As at the moment I have left the default state as on for both of those led's as it seems I can only have a one to one mapping as there is no virtual gpio function like the QSDK.

`
		yellow_mix {
			default-state = "off";
			virtuals = <red_led>, <green_led>;
		};
		blue_mix {
			default-state = "off";
			virtuals = <blue_led>, <green_led>;

professor-jonny replied:
If we could implement this then I can define LED colours direcly in the dts I believe.

 https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/drivers/leds/rgb/leds-group-multicolor.c?h=v6.10.8

professor-jonny replied:
I have just removed the script and will drive from procd, and I plan to make a kmod for the led's to support colour mixing the led's directly from DTS.

professor-jonny commented:
sorry I don't really know how to do that and I don't think that function is available with GitHub desktop.

I'm guessing I can use proper git from command line and issue a git revert command to my init commit then update my branch then force push?
Sorry I'm not very versed with git.

musashino205 commented:
Please don't use GitHub WebUI for creating commits and use local repository with Git CLI.

https://openwrt.org/docs/guide-developer/working-with-github-pr

mrkiko commented:
Hi!

First of all - thank you for your work. the device seems interesting and can attract users
Right now things will seem very hard but with persistence and patience you will be able to do it!!

As for why the preinit script is not run - be careful about the fact that if your script comes after the preinit script testing whether we are or not in initramfs, it won't be executed.
so you might want to put it before that one, with a lower starting number.

That said, in OpenWrt there's a general tendency to prefer kernel-based solution for things like LEDs, PWM fans and so on, when possible.

professor-jonny commented:
@mrkiko the device is being used by many ISP's in New Zealand as their default router for fiber with failover for 5g and I believe they are starting to use the BE65-5G which is the slightly better version with Wi-Fi 7 multiple 2.5g LAN ports and USB which I hope to get ahold of soon.

They also use the x58-4g for 4g fail over fiber service, that I have picked up yet to play with.
looking at the ISP support file embedded in the device's data partition it would seem many isp's use it all over the world.

I'd love a kernel solution for the LED's as the problem is OpenWrt is not designed to give priority of gpio-led's and there is no kernel support for multifunction/ monochromatic LED's, if two different statuses were to show at the same time the led color would be a mix not indicating the status correctly, my script was an attempt to make it useful as all the deco line of TP products have this single led or a ring of light at the top.

I don't expect my script to pass scrutineering but hopefully it sparks a conversation as it seems the multifunction monochromatic LED is all across the TP-link range of deco products.

The ability to drive a virtual LED (mix of gpio's) defined from the DTS from a single binding would be handy and make it less confusing as you can imaging to make yellow in you select reg and green colors from LUCI.

I know Prity led colors are luxury and the talented people have better things to do but I'm willing to at least make a complete hash out it and give it a go if someone can point me in the correct direction.

castillofrancodamian commented:
I think the name Deco is missing.

professor-jonny commented:
ok I can do that if deemed necessary, but would I have to do that for the board name throughout source or just the title of the commit?
I can probably change the version to be v1 too I did see that in the recommendations after a quick read just now.

castillofrancodamian commented:
In all sectors where this device is mentioned.

professor-jonny commented:
OK I fixed that I'm thinking it may break the WI-FI ? I would have to alter and resubmit the BDF with the updated ID strings?

RaylynnKnight commented:
You should take a look at [Submission Guidelines](https://openwrt.org/submitting-patches#submission_guidelines).  You are missing at least the Signed Off by line.

musashino205 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-x80-5g.dts:
```diff
@@ -0,0 +1,488 @@
+#include <dt-bindings/leds/common.h>
+
+/ {
+	model = "tplink deco-x80-5g";
```
This should be `"TP-Link Deco X80-5G"`

musashino205 commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-x80-5g.dts:
```diff
@@ -0,0 +1,488 @@
+		compatible = "gpio-leds";
+		pinctrl-0 = <&led_pins>;
+		pinctrl-names = "default";
+		// note the LED'd defined here are not individual they are a single combined RGB element
+		//to mimic the OEM default behavior green and red are switched on from boot to indicate yellow
+		//to show the device is booting behavior can be user defined in procd to override this.
+		//OEM behaviour of multi function LED:
+		//pulse Yellow: resetting/ rebooting.
+		//Solid Yellow: starting up.
+		//Pulse Blue: factory reset/ first boot ready for setup or WPS in progress.
+		//Solid Blue: setting up/ comitting settings change.
+		//Pulse Green: upgrading firmware.
+		//Solid Green: connected to 4G / WAN.
+		//Solid Purple: connected to 5G.
+		//Pulse Red: mesh failure cant find peers.
+		//Solid Red: fault
```
Please use `/* */` instead of `//` for comments.

professor-jonny replied:
Is this just for comments on their own line or does this include comments in line with code also?

musashino205 commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -332,6 +332,21 @@ define Device/redmi_ax6
+	$(call Device/FitImage)
+	$(call Device/UbiFit)
+	DEVICE_VENDOR := TP-Link
+	DEVICE_MODEL := deco-x80-5g
```
This should be `Deco X80-5G`.

robimarko replied:
Please stop marking stuff as resolved when it wasn't, and change to @musashino205 suggestion

professor-jonny replied:
I did change it from `x80-5g` to `deco-x80-5g` sorry I'm a bit lost in what is strings and what are id's but I'm learning.

robimarko commented on package/firmware/ipq-wifi/Makefile:
```diff
@@ -180,6 +181,7 @@ $(eval $(call generate-ipq-wifi-package,spectrum_sax1v1k,Spectrum SAX1V1K))
 $(eval $(call generate-ipq-wifi-package,wallys_dr40x9,Wallys DR40X9))
 $(eval $(call generate-ipq-wifi-package,xiaomi_ax3600,Xiaomi AX3600))
 $(eval $(call generate-ipq-wifi-package,xiaomi_ax9000,Xiaomi AX9000))
+$(eval $(call generate-ipq-wifi-package,tplink_deco-x80-5g,tplink deco-x80-5g))
```
The last input should be a full user visible board name, look at the others, so make that pretty.

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -332,6 +332,21 @@ define Device/redmi_ax6
+	PAGESIZE := 2048
+	DEVICE_DTS_CONFIG := config@hk01.c5
+	SOC := ipq8074
+	DEVICE_PACKAGES := kmod-hwmon-gpiofan kmod-usb3 ipq-wifi-tplink_deco-x80-5g \
```
`kmod-usb3` is already included by default

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,488 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
This comment is broken, its never closed.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,488 @@
+/* SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/* Copyright (c) 2022, Robert Marko <robimarko@gmail.com> */
```
Please look into how multiline comments work, there is no point in opening and closing comments for each line.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,488 @@
+	switch_mac_mode2 = <MAC_MODE_SGMII_CHANNEL0>; /* mac mode for uniphy instance2*/
+
+	qcom,port_phyinfo {
+		port@6 {
```
Please sort these by port index

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,488 @@
+
+	switch_lan_bmp = <ESS_PORT6>; /* lan port bitmap */
+	switch_wan_bmp = <ESS_PORT5>; /* wan port bitmap */
+	switch_mac_mode1 = <MAC_MODE_SGMII_CHANNEL0>; /* mac mode for uniphy instance1*/
```
If QCA8081 is connected here then use `MAC_MODE_SGMII_PLUS`

professor-jonny replied:
Ok thanks I have force pushed updates hopefully I have resolved all issues now

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,488 @@
+				<4855 5>, /* speed 6 binary drive bits 1,0,1 (high state) */
+				<5826 6>, /* speed 7 binary drive bits 1,1,0      ▼ */
+				<6800 7>; /* speed 8 binary drive bits 1,1,1      ▼ */
+		/* alarm-gpios = <&tlmm XX GPIO_ACTIVE_HIGH>; */
```
Drop it since it doesnt exist

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,488 @@
+				<6800 7>; /* speed 8 binary drive bits 1,1,1      ▼ */
+		/* alarm-gpios = <&tlmm XX GPIO_ACTIVE_HIGH>; */
+		cooling-min-level = <0>; /* minimum speed allowable to be set via thermal */
+		cooling-max-level = <7>; /* max speed allowable to be set via thermal */
```
I dont think `cooling-min-level` and `cooling-max-level` exist upstream, bindings dont describe them at all

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -332,6 +332,21 @@ define Device/redmi_ax6
+	SOC := ipq8074
+	DEVICE_PACKAGES := kmod-hwmon-gpiofan ipq-wifi-tplink_deco-x80-5g \
+	 	 kmod-usb-serial-option kmod-usb-net-qmi-wwan kmod-usb-serial-qualcomm uqmi\
+		 usbutils kmod-usb-storage
```
Why do you need kmod-usb-storage and usbutils by default?

professor-jonny replied:
ill drop them yip it does not need usb storage at all but the tools one was handy to diag when I was getting the modem initially peeking and poking with the modem but can be removed now.

robimarko commented:
Can you please reduce the number of comments in the DTS, all of those on okay and similar generic properties just clutter the DTS

professor-jonny commented:
Just wondering if there is something that is being waited on from my end to merge this?

robimarko commented:
It has formal isssues, that needs to be sorted out before merging.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,485 @@
+
+	qca8033: ethernet-phy@4 {
+		compatible = "ethernet-phy-ieee802.3-c22"; /* lan phy */
+		reg = <0x4>;
```
Decimal please

robimarko commented:
I merged the BDF, please update the ipq-wifi package.

robimarko commented on target/linux/qualcommax/image/ipq807x.mk:
```diff
@@ -358,6 +358,20 @@ define Device/tplink_eap660hd-v1
+	DEVICE_DTS_CONFIG := config@hk01.c5
+	SOC := ipq8074
+	DEVICE_PACKAGES := kmod-hwmon-gpiofan ipq-wifi-tplink_deco-x80-5g \
+	 	 kmod-usb-serial-option kmod-usb-net-qmi-wwan kmod-usb-serial-qualcomm uqmi
```
`kmod-usb-serial-qualcomm` should not be needed, usually, the option driver registers all of the serial interfaces.

Also, I would prefer `uqmi` to be dropped as ModemManager etc can be used for this modem as well so it makes no sense to include it by default.

robimarko commented on target/linux/qualcommax/ipq807x/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -23,4 +23,9 @@ case "$board" in
 		;;
+	tplink,deco-x80-5g)
+		base_mac=$(get_mac_binary /tmp/factory_data/default-mac 0)
+		macaddr_add $base_mac $(expr "$PHYNBR" + 1) > /sys${DEVPATH}/macaddress
```
I would prefer moving this to caldata, you can look at examples there how to patch the MAC so we dont need this fixup.

professor-jonny replied:
I tried to patch that previously, but I don't think the partition is mounted in time to patch caldata using the hotplug.d script.

I can try again

robimarko replied:
It has to be mounted since you are mounting it in preinit, otherwise caldata loading would have failed as well

professor-jonny replied:
I found the partition mounting to pull the mac binary unreliable it did not work every time.
It seems ubiatatch exits with an exitcode but the partition is not ready by the time the mound command operates, and it fails to find the MTD device. some of the time.

I added a sleep 1 in the script and it fixes the problem but the get mac binary does not work for the lan as that inits well before the partition is up and it has already got a random mac by that time.

I'm not sure how to fix the lan mac apart from writing it into bootargs in and modifying the dts.

robimarko replied:
Hm, can you try defining a fixed partition table and then using `compatible = "linux,ubi";` as that will make the kernel automount that UBI volume and that should happen really early as well.

professor-jonny replied:
ok, do I have to define the whole partition map or just the partitions I want to mount as this has an smem table?

This is the boot mountingit is after the fallback mac:
`
[    4.417036] GMAC6(ffffff800400b900) Invalid MAC@ - using fe:c0:c3:3c:71:b4
[    4.626315] init: - preinit -
[    5.060429] ubi1: attaching mtd15
[    5.110661] ubi1: scanning is finished
[    5.117811] ubi1: attached mtd15 (name "factory_data", size 9 MiB)
[    5.117852] ubi1: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    5.122891] ubi1: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    5.129781] ubi1: VID header offset: 2048 (aligned 2048), data offset: 4096
[    5.136618] ubi1: good PEBs: 72, bad PEBs: 0, corrupted PEBs: 0
[    5.143376] ubi1: user volume: 1, internal volumes: 1, max. volumes count: 128
[    5.149284] ubi1: max/mean erase counter: 16/9, WL threshold: 4096, image sequence number: 1097309032
[    5.156578] ubi1: available PEBs: 0, total reserved PEBs: 72, PEBs reserved for bad PEB handling: 20
[    5.165883] ubi1: background thread "ubi_bgt1d" started, PID 760
[    5.429679] UBIFS (ubi1:0): recovery needed
[    5.477984] UBIFS (ubi1:0): recovery deferred
[    5.478100] UBIFS (ubi1:0): UBIFS: mounted UBI device 1, volume 0, name "ubi_factory_data", R/O mode
[    5.481324] UBIFS (ubi1:0): LEB size: 126976 bytes (124 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
[    5.490548] UBIFS (ubi1:0): FS size: 4825088 bytes (4 MiB, 38 LEBs), max 48 LEBs, journal size 1142785 bytes (1 MiB, 9 LEBs)
[    5.500351] UBIFS (ubi1:0): reserved for root: 0 bytes (0 KiB)
[    5.511618] UBIFS (ubi1:0): media format: w4/r0 (latest is w5/r0), UUID 555FE482-AC27-41A3-A0C3-47CC50ED88FA, small LPT model
`

professor-jonny replied:
fixed partitions still did not fix the network mac issue but now I get the error ` OF: Bad cell count for /soc@0/nand-controller@79b0000/nand@0/partitions` during the boot log but seemingly everything still works?
is this something to do with bad blocks or do i have a bad partition map or something ?

professor-jonny replied:
found error with me baking in config during testing causing wan/ lan mac addresses to not be updated at boot.

professor-jonny commented:
I found me baking in a network config caused it to fail updating the mac addresses at boot:

https://github.com/openwrt/openwrt/issues/17391

robimarko commented on target/linux/qualcommax/ipq807x/base-files/lib/preinit/09_mount_factory_partitions:
You can roll this into the existing 09 preinit script instead of adding a new one, its per board case anyway

professor-jonny replied:
done

professor-jonny commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,593 @@
+		};
+	};
+
+	antennas {
```
You mean merge the antenna and modem nodes like so:

`
	modem {
		compatible = "gpio-export";
		pinctrl-0 = <&modem_pins>;
		pinctrl-1 = <&antenna_pins>;
		pinctrl-names = "default";

		w_disable {
			gpio-export,name = "w_disable";
			gpio-export,output = <1>;
			gpios = <&tlmm 35 GPIO_ACTIVE_HIGH>; /* wwan disable output */
		};

		modem-reset {
			gpio-export,name = "modem_reset";
			gpio-export,output = <0>;
			gpios = <&tlmm 29 GPIO_ACTIVE_HIGH>; /* reset modem output */
		};

		ext_antenna_en {
			gpio-export,name = "ext_antenna_en";
			gpio-export,output = <0>;
			gpios = <&tlmm 56 GPIO_ACTIVE_HIGH>; /* select external sma antennas for wwan output */
		};
	};
`

If this is not what you mean you might have to point me to an example, I'm guessing the #tlmn nodes define pins you just need to reference them in the node?

professor-jonny replied:
I think I have this corrected after testing changes on my device and testing the network script works on the device

musashino205 commented on target/linux/qualcommax/ipq807x/base-files/etc/board.d/02_network:
```diff
@@ -105,6 +106,11 @@ ipq807x_setup_macs()
 		label_mac=$(get_mac_binary /tmp/factory_data/default-mac 0)
 		lan_mac=$label_mac
 		;;
+		tplink,deco-x80-5g)
+			label_mac=$(get_mac_binary /tmp/factory_data/default-mac 0)
+			lan_mac=$(macaddr_add $label_mac 1)
+			wan_mac=$label_mac
+		;;
```
Please sort alphabetically and fix indentation.

robimarko commented on target/linux/qualcommax/ipq807x/base-files/lib/preinit/09_mount_factory_data:
```diff
@@ -17,4 +17,18 @@ preinit_mount_factory_data() {
 	esac
 }
 
+preinit_mount_factory_partitions() {
```
Why not simply put the factory data mounting in the existing partition (And make sure to find the path and not assume the UBI device number) and then keep runtime_data in a separate function?

professor-jonny replied:
I had issues with the partition mounting taking too long, so the mac capture did not work for Lan/Wan.


using the find_mtd_chardev does not return the correct UBI partition number it returned ubi2-0 instead or UBI2-1 or something like that so ubiatatch failed.

Hard coding the ubi worked but the atatch delay caused the mount command to fail sometimes as ubiatatch exits with stdout before the device is fully attached.

The way it is now makes it 100% reliable as before it would only work some of the time.

I could not figure out how to debug the preinit scripts by setting the preinit IP address in menuconfig.

I ended up scripting the preinit  to pipe stdout and stderr to a txt file and found introducing a wait 5 after ubuatatch fixed all the problems but it is not something id be happy with

I can relook at this, but I messed around with it for ages to get it to work.
even in electric Georges script he assumes ubi0 with the mount command and does not actually use the standard functions.

professor-jonny replied:
I have now patched the script to be universal, but I don't know how to capture the ubi device number for the mount command.

professor-jonny replied:
I have created a more universal script for mounting the factory data partition and now the smem partition table works also and seems to be reliable.

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,593 @@
+		partitions {
+
+			//compatible = "fixed-partitions"; /* define fixed partition table */
+			compatible = "qcom,smem-part"; /* define SMEM partition table */
```
Sorry but this is rather confusing, its either SMEM or fixed partitions.

In your case you are clearly specifying the whole partition table via fixed partitions

professor-jonny replied:
Yes, i did not know how to proceed here, now that I have converted back to SMEM should I remove all the fixed partitions stuff?
I did not know if there was some future use thing that it may be needed for seems like it is rather redundant now.

professor-jonny replied:
Ok I removed the comment and explained I left the partition map for future use and information but it is disabled, I hope this is ok.

robimarko replied:
Just remove the table, it makes literaly no sense to leave it there and the DTS is already full of comments

professor-jonny replied:
I have removed the partition maps now.
I have removed most of the comments much of it was me learning and noting things down.
It should be less cluttered now.

robimarko commented on target/linux/qualcommax/ipq807x/base-files/lib/preinit/09_mount_factory_data:
Doesnt it make sense to do this first and the add the device?

Cause currently, you are adding the device and then immediately modifying the script to make it universal

robimarko commented:
Also, please fix the formal issues

professor-jonny committed:
Qualcommax: IPQ807x: update preinit factory partition mount script

Patch script used for mounting of factory data partitions to not assume devices names or use hardcoded identifiers.

Signed-off-by: jonathan brophy <professor_jonny@hotmail.com>
Link: https://github.com/openwrt/openwrt/pull/16329
Signed-off-by: Robert Marko <robimarko@gmail.com>

professor-jonny committed:
Qualcommax: IPQ807x: Add support for TP-Link Deco X80-5G

The TP-link Deco X80-5G is a AX WIFI router with a single 1G (LAN) and 1 2.5G (WAN) port with a built in Quectel 5G modem.

Specifications:

Architecture	ARMv8-A (aarch64 Cortex A53, 4 cores)
Vendor	Qualcomm
Bootloader	U-Boot (2016)
System-On-Chip	Qualcomm IPQ8074A (Networking Pro 1210 platform)
CPU/Speed	2.2GHZ
Flash-Chip	ESMT F59D1G81MB-45TI 1G-bit NAND
Flash MB	128MB (1Gbit)
RAM-chip	2* ESMT M15T4G16256A-DEBG2G 2Gb DDR3L (32bit)
RAM MB	512MB (4Gbit)
WLan System-On-Chip (5g)	Qualcomm QCN5054
WLan Front end modules (5g)	8* Qorvo QPF4588
WLAN (5G)	A/N/AC/AX
WLAN Antenna (int)	8* (4 shared with 2.5G)
WLan System-On-Chip (2.4G)	Qualcomm QCN5024
WLan Front end modules (2.4G)	4* Skyworks SKY85340-11
WLAN (2.4G)	B/G/N/AX
WLAN Antenna (int)	4* (shared with 5g)
Eithernet-phy (1Gbit)	Qualcomm (Atheros) AR8033-AL1A
Ethernet (1Gbit)	1*
Eithernet-phy (2.5Gbit)	Qualcomm QCA8081
Ethernet (2.5Gbit)	1*
Switch	Embedded in the SOC?
Serial	1.8v 8n1 (testpoints)
Modem System-On-Chip	Qualcomm 5G RG50xQ (Snapdragon X55 platform)
5G modem capability	5000Mbps down 900Mbps up
5G modem bands	N1/N3/N5/N7/N8/N20/N28/N38/N40/N41/N77/N78/N79
4G modem capability	4.5G LTE-Advanced Pro (Cat20 down 2Gbps) (Cat18 up 200Mbps)
4G modem FDD bands	B1/B3/B5/B7/B8/B18/B19/B20/B26/B28
4G modem TDD bands	B34/B38/B39/B40/B41/B42/B43
Modem antenna (int)	8* (5G/4G use)
Modem antenna (ext)	2* SMA connectors (5G/4G use)
SIM type	1* Nano SIM
Telephony	1* POTS RJ11
(see wiki for more info:)
https://openwrt.org/inbox/toh/tp-link/x80-5g_v1

        Notes:
Installation Instructions:

Between antennas 7 and 8 on the PCB with the RGB harness there are three test pads labelled TP1 TP2 and TP3 connections are as follows:
  |TP2|-- RX
  |TP1|-- TX
  |TP3|-- GND

RX requires an external pullup to operate somewhere around 5 K ohm but your luck may vary the Uart is 1.8v

Set a static IP and set up a tftpserver and terminal.
power the router and quickly type the magic string “tpl” and press enter to break into u-boot
in the shell set the environment variables to enable tftp booting
setenv ipaddr (routerIP)
setenv serverip (server IP)
load you initramfs:
tftpboot 0x44000000 (serverIP):openwrt-qualcommax-ipq807x-tplink_x80-5g-initramfs-uImage.itb
boot your initramfs
bootm
upload your factory image to /tmp
format and install the factory image:
ubiformat /dev/mtd12 -y -f /tmp/openwrt-qualcommax-ipq807x-tplink_x80-5g-squashfs-factory.ubi

Note as this device swaps root partitions upon update your boot partition may be set as mtd13

Update script to mount factory tplink partition's for MAC recovery and device data.
Capture Labelmac from Factory_data partition.
Patch LAN from Labelmac.
Patch ART from Labelmac.
Set GPIO to make modem operational from boot.

GPIO fan tables work like the  stock device with three fan maps high active idle with a top fan speed of 6800 RPM.

High temp 70 deg speeds 6-8
Active temp 50 deg speeds 2-5
low temp 25 degree speed 0-1

Signed-off-by: jonathan brophy <professor_jonny@hotmail.com>
Link: https://github.com/openwrt/openwrt/pull/16329
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,469 @@
+	};
+
+	antenna_pins: antenna-state {
+		ext_antenna_en {
```
No point in the subnode, just put the contents directly under `antenna-state`

professor-jonny replied:
ok sorted now

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,469 @@
+	};
+
+	modem_pins: modem-state {
+		modem_reset {
```
No point in subnodes at all, just set `pins = "gpio29", "gpio55";` under `modem-state` as all properties are the same

robimarko commented on target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts:
```diff
@@ -0,0 +1,469 @@
+	compatible = "tplink,deco-x80-5g", "qcom,ipq8074";
+
+	aliases {
+		 serial0 = &blsp1_uart5;
```
No aliases for standard LED-s?

professor-jonny replied:
The bootloader drives the led yellow on boot to indicate power until OpenWrt takes over, to mimic the factory led color I have driven the red and green RGB elements individually with the default-state = "on";
as I can't drive multiple LEDs from a single alias.

I have however done this with my virtual color driver as it is able to group the elements to virtual led's.
I will update this target in future once I have upstreamed my driver and added it to OpenWrt, unless you have another suggestion, as without userspace scripting it is impossible to implement the OEM colors that are assigned to the standard triggers under the QSDK.

https://github.com/professor-jonny/pj_openwrt/blob/kmod_virtual_led_test/target/linux/qualcommax/files/arch/arm64/boot/dts/qcom/ipq8074-deco-x80-5g.dts#L21

robimarko replied:
Ahh, I remeber it now.

robimarko commented:
Thanks! Rebased on top of main and merged!
